### ISSUE TITLE
BUG: errors in seccomp_rule_add(3) manpage example code

### ISSUE THREAD

**Issue by valoq** [seccomp/libseccomp#57]
Using the example code in the [manpage](https://github.com/seccomp/libseccomp/blob/master/doc/man/man3/seccomp_rule_add.3) I get several errors:

> error: ‘NULL’ undeclared (first use in this function)  if (ctx == NULL)

solution: #include <stddef.h>

> error: Bad system call

solution: exit_group() and exit() are needed. whitelisting will help

working example [code](https://github.com/valoq/libseccomp/blob/master/tests/libseccompExample.c)

compiled with "gcc libeccompExample.c -lseccomp"

Now it seams to work but:
running "strace -xfc a.out" results in:

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
  0.00    0.000000           0         2           read
  0.00    0.000000           0         5           open
  0.00    0.000000           0         4           close
  0.00    0.000000           0         4           fstat
  0.00    0.000000           0         8           mmap
  0.00    0.000000           0         6           mprotect
  0.00    0.000000           0         1           munmap
  0.00    0.000000           0         4           brk
  0.00    0.000000           0         1           access
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         1           prctl
  0.00    0.000000           0         1           arch_prctl
  0.00    0.000000           0         2         1 seccomp
------ ----------- ----------- --------- --------- ----------------
100.00    0.000000                    40         1 total
```

So there is an error with the seccomp syscall?
(seccomp still seems to work)

running "strace a.out" again:

> execve("./a.out", ["./a.out"], [/* 28 vars */]) = 0
> brk(NULL)                               = 0x1941990
> access("/etc/ld.so.preload", R_OK)      = 0
> open("/etc/ld.so.preload", O_RDONLY|O_CLOEXEC) = 3
> fstat(3, {st_mode=S_IFREG|0644, st_size=0, ...}) = 0
> close(3)                                = 0
> open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
> fstat(3, {st_mode=S_IFREG|0644, st_size=118589, ...}) = 0
> mmap(NULL, 118589, PROT_READ, MAP_PRIVATE, 3, 0) = 0x3ceadf98000
> close(3)                                = 0
> open("/usr/lib/libseccomp.so.2", O_RDONLY|O_CLOEXEC) = 3
> read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\220\361\1\0\0\0\0\0"..., 832) = 832
> fstat(3, {st_mode=S_IFREG|0755, st_size=264456, ...}) = 0
> mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x3ceadf96000
> mmap(NULL, 2359552, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x3ceadb56000
> mprotect(0x3ceadb82000, 2093056, PROT_NONE) = 0
> mmap(0x3ceadd81000, 90112, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x2b000) = 0x3ceadd81000
> close(3)                                = 0
> open("/usr/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
> read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\260\3\2\0\0\0\0\0"..., 832) = 832
> fstat(3, {st_mode=S_IFREG|0755, st_size=1951744, ...}) = 0
> mmap(NULL, 3791152, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x3cead7b8000
> mprotect(0x3cead94d000, 2093056, PROT_NONE) = 0
> mmap(0x3ceadb4c000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x194000) = 0x3ceadb4c000
> mmap(0x3ceadb52000, 14640, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x3ceadb52000
> close(3)                                = 0
> mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x3ceadf94000
> arch_prctl(ARCH_SET_FS, 0x3ceadf94700)  = 0
> mprotect(0x3ceadb4c000, 16384, PROT_READ) = 0
> mprotect(0x3ceadd81000, 86016, PROT_READ) = 0
> mprotect(0x600000, 4096, PROT_READ)     = 0
> mprotect(0x3ceadfb9000, 4096, PROT_READ) = 0
> munmap(0x3ceadf98000, 118589)           = 0
> brk(NULL)                               = 0x1941990
> brk(0x1962990)                          = 0x1962990
> brk(0x1963000)                          = 0x1963000
> open("file.txt", O_RDONLY)              = 3
> prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)  = 0
> seccomp(SECCOMP_SET_MODE_STRICT, 1, NULL) = -1 EINVAL (Invalid argument)
> seccomp(SECCOMP_SET_MODE_FILTER, 0, {len=28, filter=0x1944010}) = 0
> exit_group(0)                           = ?
> +++ exited with 0 +++

we have two seccomp calls:

seccomp(SECCOMP_SET_MODE_STRICT, 1, NULL) = -1 EINVAL (Invalid argument)
seccomp(SECCOMP_SET_MODE_FILTER, 0, {len=28, filter=0x1944010}) = 0

It seems the first one fails. Why is that?

Tested on linux 3.8.11 with libseccomp 2.3.1

**Comment by pcmoore** [comment]
Thanks for the pointing out the problems in the example code!

The first seccomp(2) call is actually a little trick to see if the kernel supports the seccomp(2) syscall, if you are curious you can see commit a1f144a9a28aa1b831f7d3f481fb3e0e5e3de3aa for more information.  If you're not curious you can safely ignore this first failure as it is harmless (it's actually an invalid call).

**Comment by valoq** [comment]
Thanks for explaining.

I still feel like there should be another way to test that without failing the syscall.

**Comment by pcmoore** [comment]
Patches/suggestions are always welcome if you find a better way!

**Comment by pcmoore** [comment]
Fixed the example in f6a4e747ca16d6a7acfafe934aa4d0671b4fa70e.
